cmd/syncer/cmd/syncer.go

```diff
@@ -0,0 +1,150 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/util/sets"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/component-base/logs"
+	"k8s.io/klog/v2"
+
+	nscontroller "github.com/kcp-dev/kcp/pkg/reconciler/namespace"
+	"github.com/kcp-dev/kcp/pkg/syncer"
+)
+
+const numThreads = 2
+
+func NewSyncerCommand() *cobra.Command {
+	options := NewDefaultOptions()
+
+	syncerCommand := &cobra.Command{
+		Use:   "syncer",
+		Short: "Synchronizes resources in `kcp` assigned to the clusters",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := options.Complete(); err != nil {
+				return err
+			}
+
+			if err := options.Validate(); err != nil {
+				return err
+			}
+
+			ctx := genericapiserver.SetupSignalContext()
+			if err := options.Run(ctx); err != nil {
+				return err
+			}
+
+			<-ctx.Done()
+
+			return nil
+		},
+	}
+
+	options.BindFlags(syncerCommand.Flags())
+
+	return syncerCommand
+}
+
+func (options *Options) Validate() error {
+	if options.FromClusterName == "" {
+		return errors.New("--from-cluster is required")
+	}
+
+	return nil
+}
+
+func (options *Options) Complete() error {
+	klog.Infof("Syncing the following resource types: %s", options.SyncedResourceTypes)
+
+	// Create a client to dynamically watch "from".
+	if cfg, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: options.FromKubeconfig}, nil).ClientConfig(); err != nil {
+		return err
+	} else {
+		options.fromConfig = cfg
+	}
+
+	if options.ToKubeconfig != "" {
+		toOverrides := clientcmd.ConfigOverrides{
+			CurrentContext: options.ToContext,
+		}
+
+		if cfg, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: options.ToKubeconfig},
+			&toOverrides).ClientConfig(); err != nil {
+			return err
+		} else {
+			options.toConfig = cfg
+		}
+	}
+
+	return nil
+}
+
+func (options *Options) Run(ctx context.Context) error {
+	if err := syncer.StartSyncer(ctx, options.fromConfig, options.toConfig, sets.NewString(options.SyncedResourceTypes...), options.FromClusterName, options.PclusterID, numThreads); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type Options struct {
+	FromKubeconfig      string
+	FromClusterName     string
+	ToKubeconfig        string
+	ToContext           string
+	PclusterID          string
+	Logs                *logs.Options
+	SyncedResourceTypes []string
+
+	//non-exported members
+	toConfig   *rest.Config
+	fromConfig *rest.Config
```

Comment:
this does not belong into options.

Reply:
These are derived from the options, aren't they?

Usually we have only flags values (with very few exceptions) in option structs. Would put those into cmd/syncer/options, with AddFlags, Complete and Validate methods.

Then at the cmd/syncer/cmd level you will use the option structs from the options package and turn e.g. the kubeconfig path name into rest.Configs.